src/libsyntax/codemap.rs

```diff
@@ -132,13 +130,29 @@ pub struct Span {
     pub expn_id: ExpnId
 }
 
-/// Spans are converted to MultiSpans just before error reporting, either automatically,
-/// generated by line grouping, or manually constructed.
-/// In the latter case care should be taken to ensure that spans are ordered, disjoint,
-/// and point into the same FileMap.
+/// A collection of spans. Spans have two orthogonal attributes:
+///
+/// - they can be *primary spans*. In this case they are the locus of
+///   the error, and would be rendered with `^^^`.
+/// - they can have a *label*. In this case, the label is written next
+///   to the mark in the snippet when we render.
 #[derive(Clone)]
 pub struct MultiSpan {
-    pub spans: Vec<Span>
+    primary_spans: Vec<Span>,
+    span_labels: Vec<(Span, String)>,
+}
+
+#[derive(Clone, Debug)]
+pub struct SpanLabel {
+    /// the span we are going to include in the final snippet
+    pub span: Span,
+
+    /// is this a primary span? This is the "locus" of the message,
+    /// and is indicated with a `^^^^` underline, versus `----`
+    pub is_primary: bool,
+
+    /// what label should we attach to this span (if any)?
```

Comment:
Nit: comments should be proper sentences - start with capitals, end with periods.